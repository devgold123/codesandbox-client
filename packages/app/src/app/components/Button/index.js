import React from 'react';
import { LinkButton, AButton, Button } from './elements';

type Props = {
  [key: any]: any,
  to: ?string,
  href: ?string,
  small: ?boolean,
  style: ?any,
};

function ButtonComponent({ small = false, style = {}, ...props }: Props) {
  const newStyle = {
    ...(small
      ? {
          padding: '0.5em 0.7em',
          fontSize: '0.875em',
        }
      : {
          padding: '0.65em 2.25em',
        }),
    ...style,
  };

  // Link
  if (props.to) {
    return <LinkButton style={newStyle} {...props} />;
  }

  if (props.href) {
    return <AButton style={newStyle} {...props} />;
  }

  return <Button style={newStyle} {...props} />;
}

export default ButtonComponent;
