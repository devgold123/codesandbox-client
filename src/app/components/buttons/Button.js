import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import theme from 'common/theme';

const getBackgroundColor = ({ secondary, transparent, disabled }) => {
  if (disabled) return theme.background2.darken(0.1)();
  if (transparent) return 'rgba(0,0,0,0.0)';
  if (secondary) return theme.primary.clearer(0.8)();
  return theme.secondary.clearer(0.4)();
};

const getBorder = ({ transparent, disabled }) => {
  if (transparent) return `1px solid ${theme.secondary.clearer(0.5)()}`;
  if (disabled) return '1px solid transparent';
  return `1px solid ${theme.secondary()};`;
};

const getColor = ({ transparent, disabled }) => {
  if (disabled) return theme.background2.lighten(1.5)();
  if (transparent) return 'rgba(255,255,255,0.8)';
  return 'white';
};

const styles = props =>
  `
  transition: 0.3s ease all;
  text-transform: uppercase;
  text-decoration: none;
  line-height: 1;
  font-weight: 300;
  background-color: ${getBackgroundColor(props)};
  border: ${getBorder(props)};
  color: ${getColor(props)};
  border-radius: 2px;
  ${(() => {
    if (props.small) {
      return `
        padding: 0.5rem 0.75rem;
        font-size: 0.875rem;
      `;
    }
    return 'padding: 0.75rem 1rem;';
  })()}
  outline: none;
  width: ${props.block ? '100%' : 'inherit'};

  ${!props.disabled && `
      cursor: pointer;
      &:hover {
        background-color: ${theme.primary.clearer(0.5)()};
        border-color: ${theme.primary()};
      }
  `}
`;

const LinkButton = styled(Link)`${styles}`;
const AButton = styled.a`${styles}`;
const Button = styled.button`${styles}`;

type Props = {
  [key: any]: any,
  to: ?string,
  href: ?string,
};

export default (props: Props) => {
  // Link
  if (props.to) {
    return <LinkButton {...props} />;
  }

  if (props.href) {
    return <AButton {...props} />;
  }

  return <Button {...props} />;
};
