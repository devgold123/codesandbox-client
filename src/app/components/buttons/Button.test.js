import React from 'react';
import testRender from 'app/utils/test/render';
import { MemoryRouter } from 'react-router-dom';
import Button from './Button';

describe('Button', () => {
  it('renders', () => {
    testRender(<Button>Test</Button>);
  });

  it('renders onClick', () => {
    testRender(<Button onClick={() => {}}>Test</Button>);
  });

  it('renders hrefs', () => {
    testRender(
      <MemoryRouter>
        <Button to="https://ivesvh.com">Test</Button>
      </MemoryRouter>,
    );
  });

  it('renders properties', () => {
    testRender(<Button small>Test</Button>);
  });

  it('renders disabled', () => {
    testRender(<Button disabled>Test</Button>);
  });
});
