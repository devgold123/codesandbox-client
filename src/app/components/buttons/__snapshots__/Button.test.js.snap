// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
<button
  className="sc-htpNat hUeRGA"
>
  Test
</button>
`;

exports[`Button renders disabled 1`] = `
<button
  className="sc-htpNat izIXMc"
  disabled={true}
>
  Test
</button>
`;

exports[`Button renders href 1`] = `
<a
  className="sc-bwzfXH cMAwix"
  href="https://ivesvh.com"
>
  Test
</a>
`;

exports[`Button renders onClick 1`] = `
<button
  className="sc-htpNat hUeRGA"
  onClick={[Function]}
>
  Test
</button>
`;

exports[`Button renders properties 1`] = `
<button
  className="sc-htpNat kgEmQI"
>
  Test
</button>
`;

exports[`Button renders to 1`] = `
<a
  className="sc-bdVaJa kfQsdG"
  href="https://ivesvh.com"
  onClick={[Function]}
>
  Test
</a>
`;
